src/pycram/failure_handling.py

```diff
@@ -79,5 +82,98 @@ def perform(self):
                     raise e
 
 
+class RetryMonitor(FailureHandling):
+    """
+    A subclass of FailureHandling that implements a retry mechanism that works with a Monitor.
+
+    This class represents a specific failure handling strategy that allows us to retry a demo that is
+    being monitored, in case that monitoring condition is triggered.
+
+    Attributes:
+        max_tries (int): The maximum number of attempts to retry the action.
+        recovery (dict): A dictionary that maps exception types to recovery actions
+
+    Inherits:
+        All attributes and methods from the FailureHandling class.
 
+    Overrides:
+        perform(): Implements the retry logic.
+    """
+
+    def __init__(self, designator_description: Monitor, max_tries: int = 3, recovery: dict = None):
+        """
+        Initializes a new instance of the Retry class.
+
+        Args:
+            designator_description (DesignatorDescription): The description or context
+            of the task or process for which the retry mechanism is being set up.
+            max_tries (int, optional): The maximum number of attempts to retry. Defaults to 3.
+            recovery (dict, optional): A dictionary that maps exception types to recovery actions. Defaults to None.
+        """
+        super().__init__(designator_description)
+        self.max_tries = max_tries
+        self.lock = Lock()
+        if recovery is None:
+            self.recovery = {}
+        else:
+            if not isinstance(recovery, dict):
+                raise ValueError(
+                    "Recovery must be a dictionary with exception types as keys and Language instances as values.")
+            for key, value in recovery.items():
+                if not issubclass(key, BaseException):
+                    raise TypeError("Keys in the recovery dictionary must be exception types.")
+                if not isinstance(value, Language):
+                    raise TypeError("Values in the recovery dictionary must be instances of the Language class.")
+            self.recovery = recovery
+
+    def perform(self):
+        """
+        Implementation of the retry mechanism.
+
+        This method attempts to perform the Monitor + plan specified in the designator_description.
+        If the action fails, it is retried up to max_tries times. If all attempts fail,
+        the last exception is raised. In every loop, we need to clear the kill_event, and set all
+        relevant 'interrupted' variables to False, to make sure the Monitor and plan are executed
+        properly again
+
+        Raises:
+            PlanFailure: If all retry attempts fail.
+
+        Returns:
+            The state of the execution performed, as well as a flattened list of the results, in the correct order
+        """
 
+        def reset_interrupted(child):
+            if hasattr(child, "interrupted"):
```

Comment:
this is a very unsafe way of doing things. Why is child.interrupted not guaranteed?

Reply:
i wasnt aware that its unsafe, removed it as it should not be needed anymore. where needed im using try-except now

Reply:
Hm, try except is not the solution i was looking for. If you expect the child to be of a certain type that has an interrupted attribute you should specify that in the typing and can expect that every child has it. Then just access the attribute. If someone passed in a wrong type he would get a yellow marking of the python linter and if executed anyway an attribute error that. Is that not intended?

Reply:
Well the problem is that the type here may for example be a "Code()" language object, but due to python limitations, as i understand it, there are no ways to properly interrupt these code blocks. Thats why currently, the interrupt method of Code() raises a NotImplementedError. Also, this should never be called by hand, only through the Monitor, which can only do monitoring for Language objects. and these objects either have a interrupt method or it raises a NotImplementedError

Reply:
@tomsch420 pls answer